low-level/sspl_ll_d

```diff
@@ -544,42 +566,80 @@ def _sensors_actuators_factory(sspl_threaded_modules, msgQlist, operating_system
+                               recovery_count=0, recovery_interval=0):
+    logger.info("Starting: %s" % module.name())
 
-        # Shut down any running threads, error is non-recoverable
-        # shutdown_handler()
+    # Initialize persistence cache for sensor status
+    node_id = Conf.get(GLOBAL_CONF, NODE_ID_KEY, "SN01")
+    per_data_path = os.path.join(MODULE_CACHE_DIR,
+                f'{module.name().upper()}_{node_id}')
+    if not os.path.isfile(per_data_path):
+        module_persistent_data[module.name()] = {}
+        store.put(module_persistent_data[module.name()], per_data_path)
 
-        # Pause for threads to cleanup and kill main process
-        # logger.info("Killing main process in 10 seconds to allow thread cleanup time")
-        # time.sleep(10)
-        # logger.info("Systemd will restart SSPL-LL momentarily")
-        # os.kill(os.getpid(), signal.SIGKILL)
+    recovery_enabled = True if recovery_count else False
+    attempt = 0
 
-        #TODO: Restart thread instead of shutdown
-        logger.error(f"{curr_module.name()} has encountered an error {ex}, error is unrecoverable , shutting down {curr_module.name()}")
-        curr_module.shutdown()
+    while attempt <= recovery_count:
+        attempt += 1
+        try:
+            # Suspend module threads
+            if resume == False:
+                module.suspend()
+            # Each module is passed a reference list to message queues so it can transmit
+            # internal messages to other modules as desired
+            module.start_thread(conf_reader, msgQlist, product, recovery_enabled)
+        except Exception as err:
+            curr_state = "fault"
+            err_msg = f"{module.name()} has encountered an error: {err}. "
+            if attempt <= recovery_count:
+                logger.warn(err_msg)
+                logger.debug(f"Recovering {module.name()} from failure, "
+                             f"attempt: {attempt}")
+                time.sleep(recovery_interval)
+            else:
+                # Populate an actuator response message and transmit back to HAlon
+                # that we have a fatal error
+                error_msg = (f"{err_msg}. Error is unrecoverable, "
+                             f"shutting down {module.name()}. Please "
+                             "restart sspl service. In case issue persists "
+                             "even after restarting sspl service, please"
+                             "reach support team further.")
+                logger.error(error_msg)
+
+                # Check previous state before sending response message
+                if os.path.isfile(per_data_path):
+                    module_persistent_data[module.name()] = store.get(per_data_path)
+                prv_state = module_persistent_data[module.name()].get('prev_state')
+                if curr_state != prv_state:
+                    send_thread_controller_request
```

Comment:
Where are the parenthesis & arguments for 'send_thread_controller_request' function call. Guess tests are also pending

Reply:
I had a thought of restarting the module using ThreadController. But it cares only run method after initialization().
To cover both initialization() and run(), it won't be suitable. 

So this is removed. Now, in case of failure, module restart will be taken care with start_thread() itself.

Reply:
Done.